lib/ansible/modules/systemd.py

```diff
@@ -542,11 +542,16 @@ def main():
                 elif module.params['state'] == 'stopped':
                     if is_running_service(result['status']) or is_deactivating_service(result['status']):
                         action = 'stop'
-                else:
+                elif module.params['state'] == 'restarted':
+                    # Always explicitly restart, even if the service isn't running, to propagate that action to dependent services.
+                    # https://github.com/ansible/ansible/issues/77337
+                    action = 'restart'
+                    result['state'] = 'started'
+                else:  # 'reloaded'
                     if not is_running_service(result['status']):
                         action = 'start'
                     else:
-                        action = module.params['state'][:-2]  # remove 'ed' from restarted/reloaded
+                        action = 'reload'
```

Comment:
This PR is missing a changelog and tests, but also I don't think this is the correct approach. This module has the same problem with other commands (dependencies are not accounted for) and requires a more generic solution to handle all cases by always taking dependencies into consideration.

Sadly I have not found a good way to get this information in a usable form from systemd w/o going into dbus and that makes the solution very complicated.

https://docs.ansible.com/ansible/devel/community/development_process.html#micro-development-the-lifecycle-of-a-pr

Reply:
I've added a changelog fragment as requested. I could also add a test if you could point me to an existing systemd test for restarting a service.

While my approach may not solve all problems with dependencies, it solves the specific problem I laid out in #77337.
As you write yourself that a generic solution to solve all systemd dependency problems is likely out of reach for the foreseeable future, I would appreciate if my PR could be merged as a first step.